astra-sim/workload/HardwareResource.cc

```diff
@@ -12,45 +12,37 @@ using namespace Chakra;
 typedef ChakraProtoMsg::NodeType ChakraNodeType;
 
 HardwareResource::HardwareResource(uint32_t num_npus)
-    : num_npus(num_npus),
-      num_in_flight_mem_reqs(0),
-      num_in_flight_comps(0),
-      num_in_flight_comms(0) {}
+    : num_npus(num_npus), num_in_flight_cpu_ops(0), num_in_flight_gpu_ops(0) {}
 
 void HardwareResource::occupy(const shared_ptr<Chakra::ETFeederNode> node) {
-  if ((node->getChakraNode()->node_type() == ChakraNodeType::MEM_LOAD_NODE) ||
-      (node->getChakraNode()->node_type() == ChakraNodeType::MEM_STORE_NODE)) {
-    ++num_in_flight_mem_reqs;
-  } else if (node->getChakraNode()->node_type() == ChakraNodeType::COMP_NODE) {
-    ++num_in_flight_comps;
-  } else if (
-      (node->getChakraNode()->node_type() == ChakraNodeType::COMM_SEND_NODE) ||
-      (node->getChakraNode()->node_type() == ChakraNodeType::COMM_RECV_NODE) ||
-      (node->getChakraNode()->node_type() == ChakraNodeType::COMM_COLL_NODE)) {
-    ++num_in_flight_comms;
+  if (node->is_cpu_op()) {
+    ++num_in_flight_cpu_ops;
+  } else {
+    ++num_in_flight_gpu_ops;
   }
 }
 
 void HardwareResource::release(const shared_ptr<Chakra::ETFeederNode> node) {
-  if ((node->getChakraNode()->node_type() == ChakraNodeType::MEM_LOAD_NODE) ||
-      (node->getChakraNode()->node_type() == ChakraNodeType::MEM_STORE_NODE)) {
-    --num_in_flight_mem_reqs;
-  } else if (node->getChakraNode()->node_type() == ChakraNodeType::COMP_NODE) {
-    --num_in_flight_comps;
-  } else if (
-      (node->getChakraNode()->node_type() == ChakraNodeType::COMM_SEND_NODE) ||
-      (node->getChakraNode()->node_type() == ChakraNodeType::COMM_RECV_NODE) ||
-      (node->getChakraNode()->node_type() == ChakraNodeType::COMM_COLL_NODE)) {
-    --num_in_flight_comms;
+  if (node->is_cpu_op()) {
+    --num_in_flight_cpu_ops;
+  } else {
+    --num_in_flight_gpu_ops;
   }
 }
 
 bool HardwareResource::is_available(
     const shared_ptr<Chakra::ETFeederNode> node) const {
-  if ((node->getChakraNode()->node_type() == ChakraNodeType::COMP_NODE) &&
-      (num_in_flight_comps >= num_npus)) {
-    return false;
+  if (node->is_cpu_op()) {
+      if (num_in_flight_cpu_ops == 0) {
```

Comment:
So is this assuming there are only one CPU or one GPU available?

Reply:
from each chakra trace perspective - you have one trace per GPU and it has corresponding CPU thread.

Reply:
But you don't need a cooresponding CPU thread, correct?  There are many different ways users can create charka traces.  It is not just directly from PyTorch.

Reply:
Having a CPU op is not mandatory from my understanding.